dbcon/joblist/tuplehashjoin.cpp

```diff
@@ -200,140 +202,237 @@ void TupleHashJoinStep::join()
+        do endOfInput on the output */
+    if (cancelled())
+    {
+//		cout << "HJ stopping... status is " << status() << endl;
+        if (largeBPS)
+            largeBPS->abort();
+
+        bool more = true;
+        RGData oneRG;
+        while (more)
+            more = smallDLs[index]->next(smallIts[index], &oneRG);
+    }
+    //joiner->doneInserting();
+
+    end_time = boost::posix_time::microsec_clock::universal_time();
+    if (!(fSessionId & 0x80000000))
+        cout << "hash table construction time = " << end_time - start_time <<
+        " size = " << joiner->size() << endl;
+
+    extendedInfo += "\n";
+
     ostringstream oss;
+    if (joiner->inPM())
+    {
+        oss << "PM join (" << index << ")" << endl;
+        #ifdef JLF_DEBUG
+        cout << oss.str();
+        #endif
+        extendedInfo += oss.str();
+    }
+    else if (joiner->inUM() && !joiner->onDisk())
+    {
+        oss << "UM join (" << index << ")" << endl;
+        #ifdef JLF_DEBUG
+        cout << oss.str();
+        #endif
+        extendedInfo += oss.str();
+    }
 
-    try
+    /* Trying to get the extended info to match the original version
+    It's kind of kludgey at the moment, need to clean it up at some point */
+    if (!joiner->onDisk())
     {
-        /* check for join types unsupported on the PM. */
-        if (!largeBPS || !isExeMgr)
-        {
-            flippedUMSwitch = true;
-            oss << "UM join (" << index << ")";
-#ifdef JLF_DEBUG
-            cout << oss.str() << endl;
-#endif
-            extendedInfo += oss.str();
-            joiner->setInUM();
-        }
+        joiner->doneInserting();
+        boost::mutex::scoped_lock lk(*fStatsMutexPtr);
+        fExtendedInfo += extendedInfo;
+        formatMiniStats(index);
+    }
+}
 
-        resourceManager->getMemory(joiner->getMemUsage(), sessionMemLimit, false);
-        (void)atomicops::atomicAdd(&totalUMMemoryUsage, joiner->getMemUsage());
-        memUsedByEachJoin[index] += joiner->getMemUsage();
+/* Index is which small input to read. */
+void TupleHashJoinStep::smallRunnerFcn(uint32_t index, uint threadID, uint64_t *jobs)
+{
+    utils::setThreadName("HJSmallRunner");
+    bool more = true;
+    RGData oneRG;
+    Row r;
+    RowGroupDL* smallDL;
+    uint32_t smallIt;
+    RowGroup smallRG;
+    boost::shared_ptr<TupleJoiner> joiner = joiners[index];
+
+    smallDL = smallDLs[index];
+    smallIt = smallIts[index];
+    smallRG = smallRGs[index];
 
+    smallRG.initRow(&r);
+
+    try
+    {
+        ssize_t rgSize;
+        bool gotMem;
+        goto next;
         while (more && !cancelled())
         {
-            uint64_t memUseBefore, memUseAfter;
-
             smallRG.setData(&oneRG);
-
             if (smallRG.getRowCount() == 0)
                 goto next;
 
-            smallRG.getRow(0, &r);
-
-            memUseBefore = joiner->getMemUsage() + rgDataSize;
-
             // TupleHJ owns the row memory
+            utils::getSpinlock(rgdLock);
```

Comment:
It looks like a presumtion that no thread could be slowdown while holding this lock. is condition_variable is an overkill here?

Reply:
mutex would be the replacement.  For typically very short waits like this, a spinlock is more efficient, for anything longer, mutex is more efficient b/c it frees up CPU cycles.  In this case I benchmarked it both ways and couldn't observe a difference.  Went with the one that doesn't involve an interrupt.